crates/flipperzero/src/furi/string.rs

```diff
@@ -0,0 +1,259 @@
+//! String primitives built around `FuriString`.
+
+use core::{
+    cmp::Ordering,
+    convert::Infallible,
+    ffi::{c_char, CStr},
+    fmt, ops,
+};
+
+#[cfg(feature = "alloc")]
+use alloc::ffi::CString;
+
+use flipperzero_sys as sys;
+
+/// A Furi string.
+///
+/// This is similar to Rust's [`String`], but used inside the Flipper Zero SDK.
```

Comment:
I think it's actually closer to `CString` as it isn't guaranteed to be UTF-8 encoded and is always `nul` terminated.

Reply:
Yeah, I originally wrote that, but Rust's `CString` is a very minimal type, whereas here I'm trying to expose a similar interface to `String` (given that the Flipper Zero SDK exposes all that functionality). The `nul` termination in particular doesn't really matter in this context, because it is handled for you (vs when using a `CString` you explicitly need to care).

I guess it can be described as "This is similar to Rust's `CString`, but with the flexibility of Rust's `String`". I'll think of something to write here.

Reply:
Sort of related, I've been wondering where the best place to put a `FuriString` Rust type would be.

There's a few fundamental types, like this one, that would be good to have safe wrappers for low level use (e.g. in `flipperzero-rt`'s panic handler). So far I've been stuffing these into `flipperzero-sys`, but perhaps this warrants a `flipperzero-core` for the lowest level of safe wrappers. Then `flipperzero` could provide a wrapper type with stronger guarantees (e.g. is always valid UTF-8).

Reply:
I put this under `flipperzero::furi` because that's where the other Furi wrappers were. Happy to move it elsewhere. Another idea: a `flipperzero-furi` crate, that can then be used directly by e.g. `flipperzero-rt` and be re-exported as `flipperzero::furi`?

> Then `flipperzero` could provide a wrapper type with stronger guarantees (e.g. is always valid UTF-8).

Something I noticed while working on this PR is that `String` and `str` have diverse APIs, while `CString` and `CStr` are quite minimal. For `CStr` it makes sense: you can't treat it like a string slice because C strings must always be nul-terminated. But `CString` has confused me; I guess there hasn't been much motivation to extend its API given that most Rust programs will just use `String`.

I think the direct `FuriString` wrapper should be expressive (as I've been implementing), but we'd also want the UTF-8 wrapper to be similarly usable (perhaps even more closely aligning with Rust `String`). If we did introduce this kind of wrapper, then we'd need to make it opaque without a `Deref` impl, to ensure that the "always valid UTF-8" invariant is maintained.

Reply:
> Another idea: a `flipperzero-furi` crate, that can then be used directly by e.g. `flipperzero-rt` and be re-exported as `flipperzero::furi`?

That's not a bad idea either. I think for the time being I'll just keep using `flipperzero-sys` for these low-level wrappers (the [`windows`](https://crates.io/crates/windows) crate [does something similar](https://github.com/microsoft/windows-rs/blob/master/crates/libs/sys/src/core/mod.rs)).

> But `CString` has confused me; I guess there hasn't been much motivation to extend its API given that most Rust programs will just use `String`.

Right. I really see it mostly as an intermediary for working with C APIs. A bit of a necessary evil.

The reason I do think having a `sys::FuriString` wrapper (with/without UTF-8 guarantee) is useful here is because the device is fairly memory constrained, so people might actually care about unnecessary allocations. I like your approach of an expressive `furi::String` type with a UTF-8 guarantee since it provides a far more friendly API than `CString`, while still being quite careful about resources.

I imagine a `sys::furi::String` low-level type would feel a bit more like `Vec<c_char>` that exposed the `FuriString` API, but not much else.